src/script/conversation/ConversationVerificationStateHandler/MLS/MLSStateHandler.ts

```diff
@@ -0,0 +1,165 @@
+    const statusChanged = willChangeToDegraded({
+      conversationEntity,
+      logger: this.logger,
+    });
+    if (statusChanged) {
+      const type = VerificationMessageType.UNVERIFIED;
+      const event = EventBuilder.buildDegraded(conversationEntity, userIds, type);
+      await this.eventRepository.injectEvent(event);
+    }
+  };
+
+  /**
+   * This function checks if the conversation is degraded and if it is, it will verify it
+   * @param conversationEntity
+   * @param userIds
+   */
+  private verifyConversation = async (conversationEntity: Conversation) => {
+    this.logger.log(`Conversation ${conversationEntity.name} will be verified`);
+    const statusChanged = willChangeToVerified({conversationEntity, logger: this.logger});
+    if (statusChanged) {
+      const allVerifiedEvent = EventBuilder.buildAllVerified(conversationEntity);
+      await this.eventRepository.injectEvent(allVerifiedEvent);
+    }
+  };
+
+  /**
+   * This function returns the WireIdentity of all userDeviceEntities in a conversation, as long as they have a certificate.
+   * If the conversation has userDeviceEntities without a certificate, it will not be included in the returned array
+   *
+   */
+  private getAllUserEntitiesInConversation = async (conversation: Conversation) => {
+    if (!conversation.groupId) {
+      this.logger.error('Conversation has no groupId', conversation.name);
+      throw new Error('Conversation has no groupId');
+    }
+
+    const userEntities = conversation.getAllUserEntities();
+
+    const deviceUserPairs = userEntities
+      .flatMap(userEntity => {
+        return userEntity.devices().map(device => ({[device.id]: userEntity.qualifiedId}));
+      })
+      .reduce((acc, current) => {
+        return {...acc, ...current};
+      }, {});
+
+    const identities = await this.core.service!.e2eIdentity!.getUserDeviceEntities(
+      conversation.groupId,
+      deviceUserPairs,
+    );
+
+    return {
+      identities,
+      isResultComplete: Object.keys(deviceUserPairs).length === identities.length,
+      qualifiedIds: userEntities.map(userEntity => userEntity.qualifiedId),
+    };
+  };
+
+  private async isCertificateActiveAndValid(certificateString: string): Promise<boolean> {
+    const cert = new X509Certificate(certificateString);
+    const isValid = await cert.verify();
+    const isActive = cert.notAfter.getTime() > Date.now();
+
+    return isValid && isActive;
+  }
+
+  private async checkEpoch({groupId, epoch}: {groupId: string; epoch: number}): Promise<void> {
+    this.logger.log(`Epoch changed to ${epoch} for groupId ${groupId}`);
+    const conversationEntity = getConversationByGroupId({conversationState: this.conversationState, groupId});
+    if (!conversationEntity) {
+      this.logger.error(`Epoch changed but conversationEntity can't be found`);
+      return;
+    }
+    // We only want to process MLS conversations
+    if (!isMLSConversation(conversationEntity)) {
```

Comment:
Since only MLS conversations have groupId, let's have `getConversationByGroupId` return `MLSConversation | undefined` already, so there's no need to do extra check already 🤔